provider/provider_program_test.go

```diff
@@ -24,11 +10,84 @@ import (
 
 	"github.com/stretchr/testify/require"
 
+	"github.com/pulumi/providertest"
+	"github.com/pulumi/providertest/optproviderupgrade"
 	"github.com/pulumi/providertest/pulumitest"
+	"github.com/pulumi/providertest/pulumitest/assertpreview"
 	"github.com/pulumi/providertest/pulumitest/opttest"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
 )
 
 const providerName = "artifactory"
+const defaultBaselineVersion = "6.4.4"
+
+var programs = []string{
+	"test-programs/.DS_Store",
+	"test-programs/index_DebianRepository",
+	"test-programs/index_AlpineRepository",
+}
+
+func TestUpgradeCoverage(t *testing.T) {
+	providertest.ReportUpgradeCoverage(t)
+}
+
+type UpgradeTestOpts struct {
+	baselineVersion string
+	assertFunc      func(*testing.T, auto.PreviewResult)
+	config          map[string]string
+}
+
+func WithBaselineVersion(baselineVersion string) func(opts *UpgradeTestOpts) {
+	return func(opts *UpgradeTestOpts) {
+		opts.baselineVersion = baselineVersion
+	}
+}
+
+func WithAssertFunc(assertFunc func(*testing.T, auto.PreviewResult)) func(opts *UpgradeTestOpts) {
+	return func(opts *UpgradeTestOpts) {
+		opts.assertFunc = assertFunc
+	}
+}
+
+func WithConfig(config map[string]string) func(opts *UpgradeTestOpts) {
+	return func(opts *UpgradeTestOpts) {
+		opts.config = config
+	}
+}
+func testProviderUpgrade(t *testing.T, dir string, opts ...func(*UpgradeTestOpts)) {
+	options := &UpgradeTestOpts{}
+	for _, o := range opts {
+		o(options)
+	}
+	testProviderUpgradeWithOpts(t, dir, options.baselineVersion, options.config, options.assertFunc)
+}
+
+func testProviderUpgradeWithOpts(
+	t *testing.T, dir, baselineVersion string, config map[string]string,
+	assertFunction func(*testing.T, auto.PreviewResult),
+) {
+	if testing.Short() {
+		t.Skipf("Skipping in testing.Short() mode, assuming this is a CI run without credentials")
+	}
+	cwd, err := os.Getwd()
+	require.NoError(t, err)
+	if baselineVersion == "" {
+		baselineVersion = defaultBaselineVersion
+	}
+	test := pulumitest.NewPulumiTest(t, dir,
+		opttest.DownloadProviderVersion(providerName, baselineVersion),
+		opttest.LocalProviderPath(providerName, filepath.Join(cwd, "..", "bin")),
+	)
+	for k, v := range config {
+		test.SetConfig(k, v)
+	}
+	result := providertest.PreviewProviderUpgrade(test, providerName, baselineVersion, optproviderupgrade.DisableAttach())
+	if assertFunction != nil {
+		assertFunction(t, result)
+	} else {
+		assertpreview.HasNoReplacements(t, result)
+	}
+}
```

Comment:
This is a lot of boilerplate to add that doesn't actually get invoked. Can we record for the upgrade tests so we can actually run this (and make sure it works).

Reply:
Yeah, that's the plan but I wanted to get the basic tests working first.